pkg/operator/controller/canary/controller.go

```diff
@@ -118,9 +192,119 @@ type reconciler struct {
 	client client.Client
 }
 
-// TODO: Canary Controller Phase 2
-// Add callers for these 2 functions
-//
+func (r *reconciler) startCanaryRoutePolling(stop <-chan struct{}) error {
+	// Keep track of how many canary checks have passed
+	// so the route endpoint can be periodically cycled.
+	checkCount := 0
+
+	// Keep track of successive canary check failures
+	// for status reporting.
+	successiveFail := 0
+
+	go wait.Until(func() {
+		// Get the current canary route every iteration in case it has been modified
+		haveRoute, route, err := r.currentCanaryRoute()
+		if err != nil {
+			log.Error(err, "failed to get current canary route for canary check")
+			return
+		} else if !haveRoute {
+			log.Info("canary check route does not exist")
+			return
+		}
+		// Periodically rotate the canary route endpoint.
+		if checkCount > canaryCheckCycleCount {
+			haveService, service, err := r.currentCanaryService()
+			if err != nil {
+				log.Error(err, "failed to get canary service")
+				return
+			} else if !haveService {
+				log.Info("canary check service does not exist")
+				return
+			}
+			route, err = r.rotateRouteEndpoint(service, route)
+			if err != nil {
+				log.Error(err, "failed to rotate canary route endpoint")
+				return
+			}
+			checkCount = 0
+			// Give the router time to reload by returning here.
+			return
+		}
+
+		err = probeRouteEndpoint(route)
+		host := route.Spec.Host
+		if err != nil {
+			log.Error(err, "error performing canary route check")
+			SetCanaryRouteReachableMetric(host, false)
+			successiveFail += 1
+			// Mark the default ingress controller degraded after 5 successive canary check failures
+			if successiveFail >= canaryCheckFailureCount {
+				if err := r.setCanaryFailingStatusCondition(); err != nil {
+					log.Error(err, "error updating canary status condition")
+				}
+			}
+			return
+		}
+
+		SetCanaryRouteReachableMetric(host, true)
+		if err := r.setCanaryPassingStatusCondition(); err != nil {
+			log.Error(err, "error updating canary status condition")
+		}
+		successiveFail = 0
+		checkCount++
+	}, canaryCheckFrequency, stop)
+
+	return nil
+}
+
+func (r *reconciler) setCanaryFailingStatusCondition() error {
+	cond := operatorv1.OperatorCondition{
+		Type:    ingresscontroller.IngressControllerCanaryCheckSuccessConditionType,
+		Status:  operatorv1.ConditionFalse,
+		Reason:  "CanaryChecksRepetitiveFailures",
+		Message: "Canary route checks for the default ingress controller are failing",
```

Comment:
Typo? ".... checks [to see] if the default ingress controller is failing"?

Reply:
I don't think this is a typo?

Reply:
If you read "checks" as a verb, it makes sense.